libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,892 @@
+        }
+    }
+    return (crc);
+}
+
+// handle a device info response
+void AP_RunCam::parse_device_info(const Response& response)
+{
+    _protocol_version = ProtocolVersion(response._recv_buf[1]);
+
+    uint8_t featureLowBits = response._recv_buf[2];
+    uint8_t featureHighBits = response._recv_buf[3];
+    if (_features == 0) {
+        _features = (featureHighBits << 8) | featureLowBits;
+    }
+    _state = State::INITIALIZED;
+    gcs().send_text(MAV_SEVERITY_INFO, "RunCam device initialized, features 0x%04X\n", _features.get());
+}
+
+// wait for the RunCam device to be fully ready
+bool AP_RunCam::camera_ready()
+{
+    if (_state != State::INITIALIZING && _state != State::INITIALIZED) {
+        return true;
+    }
+    return false;
+}
+
+// error handler for OSD simulation
+void AP_RunCam::simulation_OSD_cable_failed(const Response& response)
+{
+    _waiting_device_response = false;
+    if (response._command == Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION) {
+        uint8_t operationID = response._param;
+        if (operationID == uint8_t(ConnectionOperation::RCDEVICE_PROTOCOL_5KEY_FUNCTION_CLOSE)) {
+            return;
+        }
+    }
+}
+
+// constructor for a response structure
+AP_RunCam::Response::Response(AP_RunCam* device, Command commandID, uint8_t param,
+    uint32_t timeout, uint16_t maxRetryTimes, parse_func_t parserFunc)
+    : _recv_buf(device->_recv_buf),
+    _command(commandID),
+    _max_retry_times(maxRetryTimes),
+    _timeout_ms(timeout),
+    _device(device),
+    _param(param),
+    _parser_func(parserFunc)
+{
+    _request_timestamp_ms = AP_HAL::millis();
+    _expected_response_length = get_response_length(commandID);
+}
+
+// get the length of a response
+uint8_t AP_RunCam::Response::get_response_length(const Command command)
+{
+    for (uint16_t i = 0; i < RUNCAM_NUM_EXPECTED_RESPONSES; i++) {
+        if (_expected_responses_length[i].command == command) {
+            return _expected_responses_length[i].reponse_length;
+        }
+    }
+
+    return 0;
+}
+
+// process all of the pending responses, retrying as necessary
+bool AP_RunCam::process_waiting(Response& response, uint32_t now)
+{
+    while (_waiting_response_queue.peek(response)) {
+        if (response._request_timestamp_ms != 0 && now < (response._request_timestamp_ms + response._timeout_ms)) {
+            // response still in play
+            return true;
+        }
+        else if (response._max_retry_times > 0) {
+            // response timed out, so resend
+            debug("retrying[%d] command 0x%X, op 0x%X\n", int(response._max_retry_times), int(response._command), int(response._param));
+            response._device->send_packet(response._command, response._param);
+            response._recv_response_length = 0;
+            response._request_timestamp_ms = now;
+            response._max_retry_times -= 1;
+            _waiting_response_queue.update(response);
+
+            return false;
+        } else {
```

Comment:
pointless `else`  Please audit for more of these.

Reply:
Fixed